README.md
# AnimatedContent

The `AnimatedContent` package provides a widget for animating the content transitions in a Flutter application. It offers various animation types such as slide, fade, and combinations of both. The widget allows you to animate the child widget's entry and exit with customizable duration and animation styles.

## Installation

Add the following line to your `pubspec.yaml` file:

```yaml
dependencies:
  animated_content: ^1.0.0
```

## Buy Me A Coffee ☕️

<a href="https://www.buymeacoffee.com/altaysakarya" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>

### Features

* The widget supports different animation types: `slideLeftToRight`, `slideRightToLeft`, `slideUpToDown`, `slideDownToUp`, `fadeSlideLeftToRight`, `fadeSlideRightToLeft`, `fadeUpToDown`, `fadeDownToUp`.
* Customizable animation duration.
* Supports a single child widget with animated content.
* You can arrange multiple `AnimatedContent` widgets in a sequence.
* Provides flexibility for automatic or button-triggered animations.

### Usage

Import the package:

```dart
import 'package:animated_content/animated_content.dart';
```

Wrap your content widget with the AnimatedContent widget:

```dart
AnimatedContent(
  child: YourContentWidget(),
  duration: Duration(milliseconds: 500),
  animation: AnimatedContentAnimation.slideLeftToRight,
)
```

The available animation options are:

* `slideLeftToRight`
* `slideRightToLeft`
* `slideUpToDown`
* `slideDownToUp`
* `fadeSlideLeftToRight`
* `fadeSlideRightToLeft`
* `fadeUpToDown`
* `fadeDownToUp`

You can customize the animation duration by providing a `Duration` object to the `duration` parameter.


### Example

<img src="https://github.com/altaysakarya/animated_content/blob/main/example_gif.gif?raw=true" height="400" alt="Example GIF">


```dart
class AnimatedContentExample extends StatelessWidget {
  const AnimatedContentExample({super.key});

  @override
  Widget build(BuildContext context) {
    String loremText =
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text("AnimatedContentExample")),
        body: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            children: [
              const AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                child: Text(
                  "Lorem ipsum dolor",
                  style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
                ),
              ),
              const SizedBox(height: 10),
              AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                duration: const Duration(
                  milliseconds: 1000,
                ),
                child: Text(loremText, style: const TextStyle(fontSize: 16)),
              ),
              const SizedBox(height: 10),
              AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                duration: const Duration(
                  milliseconds: 1500,
                ),
                child: Text(loremText, style: const TextStyle(fontSize: 16)),
              )
            ],
          ),
        ),
      ),
    );
  }
}

```

CHANGELOG.md
## 1.0.0

* Package created

## 1.0.1

* Add a example
* License

## 1.0.2

* Change description

lib/animated_content.dart
library animated_content;
import 'package:animated_content/configs/animated_content_animation_configurator.dart';
import 'package:flutter/material.dart';

enum AnimatedContentAnimation {
  slideLeftToRight,
  slideRightToLeft,
  slideUpToDown,
  slideDownToUp,
  fadeSlideLeftToRight,
  fadeSlideRightToLeft,
  fadeUpToDown,
  fadeDownToUp,
}

class AnimatedContent extends StatefulWidget {
  final Widget child;
  final Duration duration;
  final AnimatedContentAnimation animation;

  const AnimatedContent({
    Key? key,
    required this.child,
    this.duration = const Duration(milliseconds: 500),
    this.animation = AnimatedContentAnimation.slideLeftToRight,
  }) : super(key: key);

  @override
  _AnimatedContentState createState() => _AnimatedContentState();
}

class _AnimatedContentState extends State<AnimatedContent> with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<Offset> _animation;
  late Curve _curve;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: widget.duration,
    );

    // Configure the animation based on the selected animation type
    _configureAnimation();

    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(AnimatedContent oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.animation != oldWidget.animation) {
      _configureAnimation();
      _animationController.forward(from: 0.0);
    }
  }

  void _configureAnimation() {
    // Use a separate configuration class to define animation settings
    final animationConfig = AnimatedContentAnimationConfigurator.getAnimationConfig(widget.animation);

    _animation = Tween<Offset>(
      begin: animationConfig.begin,
      end: animationConfig.end,
    ).animate(_animationController);
    _curve = animationConfig.curve;
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (BuildContext context, Widget? child) {
        return SlideTransition(
          position: _animation,
          child: _shouldApplyFadeTransition()
              ? FadeTransition(
                  opacity: _animationController.drive(CurveTween(curve: _curve)),
                  child: widget.child,
                )
              : widget.child,
        );
      },
    );
  }

  bool _shouldApplyFadeTransition() {
    final animationType = widget.animation.toString().split(".").last.substring(0, 4).toLowerCase();
    return animationType == "fade";
  }
}





lib/configs/animated_content_animation_configurator.dart
// Animation configuration class to define animation settings
import 'package:animated_content/animated_content.dart';
import 'package:animated_content/configs/animation_config.dart';
import 'package:flutter/material.dart';

class AnimatedContentAnimationConfigurator {
  static AnimationConfig getAnimationConfig(AnimatedContentAnimation animation) {
    switch (animation) {
      case AnimatedContentAnimation.slideLeftToRight:
      case AnimatedContentAnimation.fadeSlideLeftToRight:
        return AnimationConfig(
          begin: const Offset(-1.0, 0.0),
          end: Offset.zero,
          curve: Curves.easeInOut,
        );
      case AnimatedContentAnimation.slideRightToLeft:
      case AnimatedContentAnimation.fadeSlideRightToLeft:
        return AnimationConfig(
          begin: const Offset(1.0, 0.0),
          end: Offset.zero,
          curve: Curves.easeInOut,
        );
      case AnimatedContentAnimation.slideUpToDown:
      case AnimatedContentAnimation.fadeUpToDown:
        return AnimationConfig(
          begin: const Offset(0.0, -1.0),
          end: Offset.zero,
          curve: Curves.easeInOut,
        );
      case AnimatedContentAnimation.slideDownToUp:
      case AnimatedContentAnimation.fadeDownToUp:
        return AnimationConfig(
          begin: const Offset(0.0, 1.0),
          end: Offset.zero,
          curve: Curves.easeInOut,
        );
    }
  }
}

example/main.dart
import 'package:animated_content/animated_content.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const AnimatedContentExample());
}

class AnimatedContentExample extends StatelessWidget {
  const AnimatedContentExample({super.key});

  @override
  Widget build(BuildContext context) {
    String loremText =
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text("AnimatedContentExample")),
        body: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            children: [
              const AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                child: Text(
                  "Lorem ipsum dolor",
                  style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
                ),
              ),
              const SizedBox(height: 10),
              AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                duration: const Duration(
                  milliseconds: 1000,
                ),
                child: Text(loremText, style: const TextStyle(fontSize: 16)),
              ),
              const SizedBox(height: 10),
              AnimatedContent(
                animation: AnimatedContentAnimation.fadeDownToUp,
                duration: const Duration(
                  milliseconds: 1500,
                ),
                child: Text(loremText, style: const TextStyle(fontSize: 16)),
              )
            ],
          ),
        ),
      ),
    );
  }
}
